pkg/agent/route/route.go

```diff
@@ -0,0 +1,304 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package route
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"strings"
+
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/config"
+	"github.com/vmware-tanzu/antrea/pkg/agent/iptables"
+	"github.com/vmware-tanzu/antrea/pkg/agent/types"
+	"github.com/vmware-tanzu/antrea/pkg/agent/util"
+)
+
+const (
+	// AntreaServiceTable is route table name for Antrea service traffic.
+	AntreaServiceTable = "Antrea-service"
+	// AntreaServiceTableIdx is route table index for Antrea service traffic.
+	AntreaServiceTableIdx = 300
+	routeTableConfigPath  = "/etc/iproute2/rt_tables"
+	AntreaIPRulePriority  = 300
+)
+
+// Client is route client.
+type Client struct {
+	nodeConfig *types.NodeConfig
+	encapMode  config.TrafficEncapModeType
+}
+
+type serviceRtTableConfig struct {
+	Idx  int
+	Name string
+}
+
+func (s *serviceRtTableConfig) String() string {
+	return fmt.Sprintf("%s: idx %d", s.Name, s.Idx)
+}
+
+func (s *serviceRtTableConfig) IsMainTable() bool {
+	return s.Name == "main"
+}
+
+var (
+	// ServiceRtTable contains Antrea service route table information.
+	ServiceRtTable = &serviceRtTableConfig{Idx: 254, Name: "main"}
+)
+
+// NewClient returns a route client
+func NewClient() *Client {
+	return &Client{}
+}
+
+// Initialize sets up route tables for Antrea.
+func (c *Client) Initialize(nodeConfig *types.NodeConfig, encapMode config.TrafficEncapModeType) error {
+	c.nodeConfig = nodeConfig
+	c.encapMode = encapMode
+	if c.encapMode.SupportsNoEncap() {
+		ServiceRtTable.Idx = AntreaServiceTableIdx
+		ServiceRtTable.Name = AntreaServiceTable
+	}
+
+	if ServiceRtTable.IsMainTable() {
+		_ = c.RemoveServiceRouting()
+		return nil
+	}
+	f, err := os.OpenFile(routeTableConfigPath, os.O_RDWR|os.O_APPEND, 0)
+	if err != nil {
+		klog.Fatalf("Unable to create service route table(open):  %v", err)
+	}
+	defer func() { _ = f.Close() }()
+
+	oldTablesRaw := make([]byte, 1024)
+	bLen, err := f.Read(oldTablesRaw)
+	if err != nil {
+		klog.Fatalf("Unable to create service route table(read): %v", err)
+	}
+	oldTables := string(oldTablesRaw[:bLen])
+	newTable := fmt.Sprintf("%d %s", ServiceRtTable.Idx, ServiceRtTable.Name)
+
+	if strings.Index(oldTables, newTable) == -1 {
+		if _, err := f.WriteString(newTable); err != nil {
+			klog.Fatalf("Failed to add antrea service route table: %v", err)
+		}
+	}
+
+	gwConfig := c.nodeConfig.GatewayConfig
+	if gwConfig != nil && c.nodeConfig.PodCIDR != nil {
+		// add local podCIDR if applicable to service rt table
```

Comment:
Nit: add -> Add. Add '.' at end of the line.

Reply:
done